File: backend/app/models/lab.py
"""
SQLAlchemy ORM model for the ``labs`` table.

Lab results track ordered medical tests for a patient through a
status lifecycle: ``ordered`` → ``in_progress`` → ``completed``.

Fields include the test name, ordered date, current status, optional
result text, result date, and free-text notes. Each lab belongs to
one patient and is cascade-deleted when the patient is removed.
"""

import uuid

from sqlalchemy import Column, DateTime, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.database import Base


class Lab(Base):
    __tablename__ = "labs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id = Column(
        UUID(as_uuid=True),
        ForeignKey("patients.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    test_name = Column(String(255), nullable=False)
    ordered_date = Column(DateTime(timezone=True), server_default=func.now())
    status = Column(String(20), nullable=False, default="ordered")
    result = Column(Text)
    result_date = Column(DateTime(timezone=True))
    notes = Column(Text)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    patient = relationship("Patient", back_populates="labs")

File: frontend/src/types/lab.ts
/**
 * TypeScript interfaces for lab result data.
 *
 * - {@link Lab} — Full lab record as returned by the API.
 * - {@link LabCreate} — Payload for ordering a new lab test.
 * - {@link LabUpdate} — Partial payload for updating lab status/results.
 * - {@link LabStatus} — Union type of valid lab statuses:
 *   `"ordered"` | `"in_progress"` | `"completed"`.
 */

export type LabStatus = "ordered" | "in_progress" | "completed";

export interface Lab {
  id: string;
  patient_id: string;
  test_name: string;
  ordered_date: string;
  status: LabStatus;
  result: string | null;
  result_date: string | null;
  notes: string | null;
  created_at: string;
}

export interface LabCreate {
  test_name: string;
  ordered_date?: string;
  status?: LabStatus;
  result?: string | null;
  result_date?: string | null;
  notes?: string | null;
}

export interface LabUpdate {
  test_name?: string;
  status?: LabStatus;
  result?: string | null;
  result_date?: string | null;
  notes?: string | null;
}
